diff --git a/src/http/ngx_http.c b/src/http/ngx_http.c
index d835f89..008c893 100644
--- a/src/http/ngx_http.c
+++ b/src/http/ngx_http.c
@@ -1237,6 +1237,7 @@ ngx_http_add_addresses(ngx_conf_t *cf, ngx_http_core_srv_conf_t *cscf,
     ngx_http_conf_addr_t  *addr;
 #if (NGX_HTTP_SSL)
     ngx_uint_t             ssl;
+    ngx_uint_t             https_allow_http;
 #endif
 #if (NGX_HTTP_V2)
     ngx_uint_t             http2;
@@ -1279,6 +1280,7 @@ ngx_http_add_addresses(ngx_conf_t *cf, ngx_http_core_srv_conf_t *cscf,
         ssl = lsopt->ssl || addr[i].opt.ssl;
         protocols |= lsopt->ssl << 1;
         protocols_prev |= addr[i].opt.ssl << 1;
+        https_allow_http = lsopt->https_allow_http || addr[i].opt.https_allow_http;
 #endif
 #if (NGX_HTTP_V2)
         http2 = lsopt->http2 || addr[i].opt.http2;
@@ -1371,6 +1373,7 @@ ngx_http_add_addresses(ngx_conf_t *cf, ngx_http_core_srv_conf_t *cscf,
         addr[i].opt.proxy_protocol = proxy_protocol;
 #if (NGX_HTTP_SSL)
         addr[i].opt.ssl = ssl;
+        addr[i].opt.https_allow_http = https_allow_http;
 #endif
 #if (NGX_HTTP_V2)
         addr[i].opt.http2 = http2;
@@ -1914,6 +1917,7 @@ ngx_http_add_addrs(ngx_conf_t *cf, ngx_http_port_t *hport,
         addrs[i].conf.default_server = addr[i].default_server;
 #if (NGX_HTTP_SSL)
         addrs[i].conf.ssl = addr[i].opt.ssl;
+        addrs[i].conf.https_allow_http = addr[i].opt.https_allow_http;
 #endif
 #if (NGX_HTTP_V2)
         addrs[i].conf.http2 = addr[i].opt.http2;
@@ -1982,6 +1986,7 @@ ngx_http_add_addrs6(ngx_conf_t *cf, ngx_http_port_t *hport,
         addrs6[i].conf.default_server = addr[i].default_server;
 #if (NGX_HTTP_SSL)
         addrs6[i].conf.ssl = addr[i].opt.ssl;
+        addrs6[i].conf.https_allow_http = addr[i].opt.https_allow_http;
 #endif
 #if (NGX_HTTP_V2)
         addrs6[i].conf.http2 = addr[i].opt.http2;
diff --git a/src/http/ngx_http_core_module.c b/src/http/ngx_http_core_module.c
index 7845f8f..e4432ff 100644
--- a/src/http/ngx_http_core_module.c
+++ b/src/http/ngx_http_core_module.c
@@ -4308,6 +4308,18 @@ ngx_http_core_listen(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
             continue;
         }
 
+#if (NGX_HTTP_SSL)
+        if (ngx_strcmp(value[n].data, "https_allow_http") == 0) {
+            lsopt.https_allow_http = 1;
+            continue;
+
+            ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
+                               "the \"https_allow_http\" parameter requires "
+                               "ngx_http_ssl_module");
+            return NGX_CONF_ERROR;
+        }
+#endif
+
         ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
                            "invalid parameter \"%V\"", &value[n]);
         return NGX_CONF_ERROR;
diff --git a/src/http/ngx_http_core_module.h b/src/http/ngx_http_core_module.h
index 765e7ff..bf2c002 100644
--- a/src/http/ngx_http_core_module.h
+++ b/src/http/ngx_http_core_module.h
@@ -83,6 +83,9 @@ typedef struct {
     unsigned                   reuseport:1;
     unsigned                   so_keepalive:2;
     unsigned                   proxy_protocol:1;
+#if (NGX_HTTP_SSL)
+    unsigned                   https_allow_http:1;
+#endif
 
     int                        backlog;
     int                        rcvbuf;
@@ -241,6 +244,9 @@ struct ngx_http_addr_conf_s {
     unsigned                   http2:1;
     unsigned                   quic:1;
     unsigned                   proxy_protocol:1;
+#if (NGX_HTTP_SSL)
+    unsigned                   https_allow_http:1;
+#endif
 };
 
 
diff --git a/src/http/ngx_http_request.c b/src/http/ngx_http_request.c
index bd2be5e..d359cc4 100644
--- a/src/http/ngx_http_request.c
+++ b/src/http/ngx_http_request.c
@@ -799,6 +799,12 @@ ngx_http_ssl_handshake(ngx_event_t *rev)
 
         c->log->action = "waiting for request";
 
+#if (NGX_HTTP_SSL)
+        if (hc->addr_conf->https_allow_http) {
+            hc->ssl = 0;
+        }
+#endif
+
         rev->handler = ngx_http_wait_request_handler;
         ngx_http_wait_request_handler(rev);
 
